### ISSUE TITLE
tslint Language service plugin for TypeScript 2.2.2

### ISSUE THREAD

**Issue by angelozerr** [Microsoft/vscode-tslint#171]
Just for your information, I'm developping a tslint Language service plugin for TypeScript 2.2.1 at https://github.com/angelozerr/tslint-language-service by following the same idea than @angular/language-service.

In other words I have no double parsing of ts files (I use the SourceFile of the tsserver).

I have reused some pieces of code of your vscode-tslint (hope it will be OK, tslint-language-service is MIT license too) and adapt the code to work with TypeScript CodeFix.

I will be happy if we could work together.

Regard's Angelo

**Comment by egamma** [comment]
Hi Angelo, I was hoping this is going to happen and I'll look forward to collaborate with you on this effort!

**Comment by serkanyersen** [comment]
@angelozerr Does this mean I can drop tslint plugin and typescript itself will handle linting?

**Comment by angelozerr** [comment]
> Hi Angelo, I was hoping this is going to happen and I'll look forward to collaborate with you on this effort!

It's a great news, thanks!

> Does this mean I can drop tslint plugin and typescript itself will handle linting?

For the moment I'm consuming tslint language service inside Eclipse with [tsserver-plugins](https://github.com/angelozerr/tsserver-plugins) (so it can work with any version of TypeScript), but I will support soon TypeScript 2.2.1 (without tsserver-plugins) which provides support for plugins.

But I'm afraid that TypeScript 2.2.1 will not provide tslint codefix because of https://github.com/Microsoft/TypeScript/issues/13435

More I have not implemented the full features provided by vscode-tslint. 

In other words, please let me time to finish my work before trying tslint language service.

**Comment by serkanyersen** [comment]
Thanks for the reply. Looking forward to this.

**Comment by egamma** [comment]
@angelozerr nice work 🌹. This will not only avoid that source files are analysed by both ts and tslint, this will also support tslint rules that require a type checker. vscode-tslint currently doesn't support this (see #70).

I imagine that vscode-tslint supports a setting to use the tslint-language-service plugin instead of the tslint-language server. This would allow users to opt-in to the new support and provide feedback while not all of the current features are supported.

Here are some of the bigger issues that need to be solved from the perspective of VS Code.
- **installation** how is the tslint-language-service plugin acquired and installed. VS Code bundles the TypeScript extension which includes the Typescript server. Users should not modify the installed bits, therefore they cannot install the tslint-language-service plugin near the Typescript server. 
- **tslint versions** vscode-tslint supports that users can use different versions of tslint. They can install tslint globally or locally to the workspace and the extension will pick it up. We should continue to support this. Currently the tslint-language-service includes the tslint module as a dependency.
- **settings** settings for tslint like `configFile` or `rulesDirectory` are currently stored in user or workspace settings. I assume that these settings will be defined inside the `tsconfig.json` and no longer in user settings. There needs to be some migration.
- **auto fix** all the code fixing will eventually be done by the language-service plugin, but vscode-tslint also provides support for auto fixing on save, which we should preserve. This requires that the language-service plugin is informed when a file is saved.
- **linting on save only** typescript diagnostics are reported as the user types, for linting it can be desired to only lint on save to reduce the noise. 

I have some more feedback on the current implementation of the tslint language service in https://github.com/angelozerr/tslint-language-service/issues/6.

**Comment by angelozerr** [comment]
> @angelozerr nice work

Many thanks! But to be honnest with you, I have just tried to adapt you vscode-tslint great work to support tsserver plugins.

> This will not only avoid that source files are analysed by both ts and tslint, this will also support tslint rules that require a type checker. 

Exactly!

> vscode-tslint currently doesn't support this (see #70).

Glad tslint language service could fix this issue.

> I imagine that vscode-tslint supports a setting to use the tslint-language-service plugin instead of the tslint-language server. This would allow users to opt-in to the new support and provide feedback while not all of the current features are supported.

Yes sure.

> Here are some of the issues that need to be solved from the perspective of VS Code.

> installation how is the tslint-language-service plugin acquired and installed. VS Code bundles the TypeScript extension which includes the Typescript server. Users should not modify the installed bits, therefore they cannot install the tslint-language-service plugin near the Typescript server. Users could use the typescript.tsdk setting to use a typescript version with the tslint-language-service plug-in, but this is yet another configuration step. This is not only an issue for the tslint-language-service plugin, but for any Typescript server extensions.

I think the main problem with TypeScript 2.2.1 is the load of the plugin. It requires that tslint-language-service is hosted inside the node_modules folder where tsconfg.json is hosted too. With tsserver you have not this restriction (it uses require).
 
> tslint versions vscode-tslint supports that users can use different versions of tslint. They can install tslint globally or locally to the workspace and the extension will pick it up. We should continue to support this. Currently the tslint-language-service includes the tslint module as a direct dependency.

peer dependency should support that

> settings tslint has settings like configFile or rulesDirectory. These are currently stored in user or workspace settings. I assume that these settings will eventually be defined inside the tsconfig.json and no longer in user settings. 

It is not done, but I think it's the idea of TypeScript plugins. 

> There needs to be some migration support provided for existing users of the vscode-tslint extension.
auto fix all the code fixing will eventually be done by the language-service plugin, but vscode-tslint also provides support for auto fixing on save, which we should preserve. This requires that the language-service plugin is informed when a file is saved.

I don't know ho to do that?

> linting on save only Typescript diagnostics are reported as the user types, for linting it can be desired to only lint on save to reduce the noise.

> I have some more feedback on the current implementation of the tslint language service in angelozerr/tslint-language-service#6.

Thanks, I have tried to give you answer.

**Comment by Istar-Eldritch** [comment]
@angelozerr I'm really looking forward to this! If you need a hand / want to delegate some tasks do not doubt and ping me.

**Comment by angelozerr** [comment]
> If you need a hand / want to delegate some tasks do not doubt and ping me.

Your help is very welcome @Istar-Eldritch ! Please continue our discussion  at https://github.com/angelozerr/tslint-language-service/issues to avoid disturbing vscode-tslint project.

**Comment by angelozerr** [comment]
Hi all,

For your info, I have published https://github.com/angelozerr/tslint-language-service 0.5.0 and it starts working well with TypeScript >= 2.2.1. It provides : 

 * `tslint semantics errors`.
 * `tslint code fix`.

I have tried it with Eclipse and VSCode. The semantic errors works well with VSCode, but not the codefix. It seems that getCodeActions is not called in some case (for instance if you type only just spaces).

To use it, you must consume TypeScript >= 2.2.1 and do

`npm install tslint-language-service`

For the moment `tslint-language-service` doesn't provide the same impressive feature than vscode-tsslint. I will work on it, please create issues and any contribution and feedback are welcome! Thanks!

Thanks too to @Istar-Eldritch to have helped to improve `tslint-language-service`

**Comment by egamma** [comment]
@angelozerr I did some experiments with the tslint-language-service. I was able to get both diagnostics and code fixes to work inside vscode. 

![tslint-language-service](https://cloud.githubusercontent.com/assets/172399/25097471/f4b68d98-23a4-11e7-9849-9069f1923cc7.gif)

**Notice** this only works with TS 2.2.2 it doesn't work with TS 2.3.*, I haven't yet tracked down why.

For my experiments I've created this fork: https://github.com/egamma/tslint-language-service

I made the following changes:
- adapted the plugin creation/initialization to the plugin sample from the TS team: https://github.com/RyanCavanaugh/sample-ts-plugin. When following this approach it works out of the box for TypeScript 2.2.2.
- did some name changes to align the code more with the sample.
- following the sample the code imports the `../node_modules/typescript/lib/tsserverlibrary` instead `typescript` (there are some differences..., need to clarify this with the TS team).
- moved all the code into index.ts to follow the style of the sample, which in retrospect was not a good idea since it complicates the diffing, but you can still compare the two files `index.ts` and `ts_plugin.ts`. Sorry about that...
- I've removed the 'custom quick fixes'. While vscode-tslint has started this practice to add custom quick fixes. However, it is not the right approach, quick fixes should come from the tslint rule providers and not from tslint consumers. If there is a quick fix missing then the right way to add it is to make a pull request against the tslint repository.
- Since there is no API for registering a code fix. I did the trick to "reuse" an existing TS error code  which is registered for quick fixes for tslint warnings. I'll file an API request.
- Added some support for tslint for tslint version 5 (the fact that the code is duplicated is pain, but the duplication is not painful enough yet to come up with common abstractions to enable code sharing).
- I've removed some unused code.

**Comment by egamma** [comment]
One issue in vscode is that the user can no longer distinguish between a diagnostic from tslint and a diagnostic from TS. I've filed Microsoft/TypeScript#15224.

**Comment by angelozerr** [comment]
@egamma nice work! Is there any chance to have a PR?

**Comment by angelozerr** [comment]
> adapted the plugin creation/initialization to the plugin sample from the TS team: https://github.com/RyanCavanaugh/sample-ts-plugin. When following this approach it works out of the box for TypeScript 2.2.2.

Thanks for this sample, I didn't know. Very cool.

> did some name changes to align the code more with the sample.

> but you can still compare the two files index.ts and ts_plugin.ts. > > > Sorry about that..
No pb, your code is simple and easy to understand.

> I've removed the 'custom quick fixes'. While vscode-tslint has started this practice to add custom quick fixes. However, it is not the right approach, quick fixes should come from the tslint rule providers and not from tslint consumers. If there is a quick fix missing then the right way to add it is to make a pull request against the tslint repository.

I'm not sure with that. For instance I have added a custom quick fix "Open tslint.json" to open `tslint.json` where rule is declared. I would like to in the future open the tslint.json and go to the declared rule (no possible for the moment with tsserver which transforms location). I find this quick fix very helpful, but I'm not sure that tslint project want to provide a quick fix which does "go to rule" which fixes nothing.

> Since there is no API for registering a code fix. I did the trick to "reuse" an existing TS error code which is registered for quick fixes for tslint warnings. I've filed Microsoft/TypeScript#15227.

Yes I had created an issue too at https://github.com/Microsoft/TypeScript/issues/13435

> Added some support for tslint for tslint version 5 (the fact that the code is duplicated is pain, but the duplication is not painful enough yet to come up with common abstractions to enable code sharing).

Ok

> I've removed some unused code.

Ok

If it's possible please create a PR, thanks.

**Comment by egamma** [comment]
@angelozerr I will definitly do a PR.

Regarding custom quick fixes, I mean a custom quick fix for fixing a tslint rule when tslint doesn't provide a quick fix. For example in the current code there is a custom quick fix for the "no-var-keyword" rule, such a quick fix this is better provided by tslint. I fully agree that 'helper' code actions like 'open the tslint.json', 'apply all fixes for the same rule' make perfect sense and we need to investigate how to provide them (e.g. implement them in the server, send a custom message to the IDE, etc.).

I've closed Microsoft/TypeScript#15227 as a duplicate of Microsoft/TypeScript#13435

**Comment by angelozerr** [comment]
> @angelozerr I will definitly do a PR.

Thanks! I think to register a custom codefix, you can use my uggly code https://github.com/angelozerr/tslint-language-service/blob/master/src/ts_plugin.ts#L118

> Regarding custom quick fixes, I mean a custom quick fix for fixing a tslint rule when tslint doesn't provide a quick fix

> I fully agree that 'helper' code actions like 'open the tslint.json', 

It already exists with my cuurent code, but I cannot select the rule because of tsserver which recomputes location.

> 'apply all fixes for the same rule' make perfect sense and we need to investigate how to provide them

Yes it's my scope.

It's perfect we are totally agree, cool!

**Comment by egamma** [comment]
@angelozerr PR is submitted.

>I think to register a custom codefix, you can use my uggly code https://github.com/angelozerr/tslint-language-service/blob/master/src/ts_plugin.ts#L118

I defer this to you, but I haven't found a downside in reusing an existing TS error code. Both the quick fixes for TS and the ones for tslint appear and seem to work.

**Comment by angelozerr** [comment]
@egamma I have approved your PR and I have improved it to use custom error code and add "Open tslint.json" quick fix.

**Comment by egamma** [comment]
Closing this issue. The work on this is now tracked in the https://github.com/angelozerr/tslint-language-service repository. A version that works in VS Code and Eclipse with tslint 4/5 is now available pls see https://github.com/angelozerr/tslint-language-service/blob/master/README.md
